### PR TITLE
Binary file format

### PR DESCRIPTION
**Pull request by OmnipotentEntity**
This is a first attempt at defining a binary file format.

I fully expect something to be wrong, and I believe that in fact there is something already wrong with the value net output (though the policy net seems fine at least), though I cannot locate what I've done wrong.  So if someone who is more comfortable with the format of the network can chime in that would be helpful.

As mentioned in #1733, this is not enough still.  Because neither the training pipeline nor the server yet has support for this file format.  But this should be enough to get started on those.

The latest 20b network is only 48042715 bytes (46MB) when uncompressed.  It does compress a bit further with gzip, down to 43MB.  This is under half of the gzip compressed file on the server, so this is a big gain in that area at the very least.

### CODE REVIEW

File: src/Network.cpp [hunk 1]
```diff
@@ -267,6 +267,334 @@ std::pair<int, int> Network::load_v1_network(std::istream& wtfile) {
...
+    };
+
+    for (int block=0; block < 1 + (2 * blocks); ++block) {
+        // Convolution Weights
+        {
+            int count = filters * filters * 9;
+            if (!block) { // Very first has a different shape because it's the input layer
+                count = filters * 162;
+            }
+            std::vector<float> weights = process(count);
```
Comment (ihavnoid) [inline src/Network.cpp]: Suggest putting
```
auto weights = process(count);
```
and something similar for rest of the process() calls.

File: src/Network.cpp [hunk 2]
```diff
@@ -267,6 +267,334 @@ std::pair<int, int> Network::load_v1_network(std::istream& wtfile) {
...
+    }
+
+    // Size 1
+    {
+        std::vector<float> weights = process(m_ip2_val_b.size());
+        if (!weights.size()) {
+            return {0, 0};
+        }
+        std::copy(cbegin(weights), cend(weights), begin(m_ip2_val_b));
+    }
```
Comment (ihavnoid) [inline src/Network.cpp]: This looks like a lot of repetitive code.  How about moving the code body into process() calls and handle the ```return {0,0};``` case as an exception throw?

**Comment by ihavnoid** [comment]
I like the implementation - but suggest using ```auto``` whenever possible, for consistency with rest of the leela-zero code.

**Comment by Ttl** [comment]
The output with #1736 is 52 MB compressed. If the space saving from binary is just 9 MB I'm not sure if it's worth the trouble.

**Comment by ihavnoid** [comment]
@Ttl I guess we have to see the size of the compressed binary format.  Plus, this is without losing any precision, and this code actually includes half-precision binary formats too.  My hope is that half precision weight will further save area while not losing much strength, but that will need some experimenting.

One question that I had was whether it is okay to lose precision on the final fully connected layers (when half precision is used to store weights).  The previous half precision implementations used half precision only within the residual tower so it isn't totally equivalent.  I would like to make sure that half-precision binary weights are okay, or at least keep the final few layers to be single precision even when half precision format is used.

**Comment by ihavnoid** [comment]
Okay perhaps I misunderstood - the 48MB numbers are for the fp16 type.  So yes probably we aren't saving a lot of disk.  But anyway will probably need to see the strength difference.

**Comment by OmnipotentEntity** [comment]
@Ttl I could also attempt some form of value length encoding to save more space.  The reason why the space saving seems less than it ought to is because we're already not storing a lot of entropy.  Though I'd argue that 9 MB on 52MB is still significant, as it's around 18%.

**Comment by OmnipotentEntity** [comment]
This should be everything outside of the training code, which I will be the first to admit I do not fully understand and do not really have a good method of testing.  May I get a review?  Is the file format sane and fully featured?  Does the code make sense and is it readable?  Is this something that the project is interested in?

**Comment by ihavnoid** [comment]
My opinion is that we should eventually migrate to some form of binary format.  I believe the binary format should be at least a little bit faster than the text format (since we don't need any parsing at all) if done properly.  The current text files aren't that useful for debugging anyway (large files with very long lines - too hard to do anything on a text editor) but this may be just me.

The reason I say "eventually" is because changing file formats can cause quite a bit of trouble - we have to force everybody to move beyond a certain version (which we sort of do when forcing a minimum autogtp/leelaz version for training) but I am not sure what else we are breaking...

**Comment by OmnipotentEntity** [comment]
Did a little bit of performance tuning:

Type (20x256) | Seconds
------------ | -------------
v3 | 0.731481
v3.gz | 0.652128
v1 | 1.35178
v1.gz | 1.13941

**Comment by ihavnoid** [comment]
@OmnipotentEntity - maybe not 'breaking', but at least this will make things be a bit more complicated.  I think we should still keep the 32-bit formats for training, but only the distribution versions of the net will be binary-converted to 16-bit.  This adds a bit more complexity.

Other than that, my only other concern is that we also reduced precision for the final 'dual head' part of the net.  Our current half precision net implementation applies half precision on the residual towers, and the 'dual head' is still single precision.  I see that the final dual-head parameters have quite big numbers (~70) and on some older nets I recall seeing values in the range of thousands.  My suggestion is that we should be a bit more conservative (after all, the 'dual head' is only a very small part of the weights) unless we have proof that it is okay to use fp16 even on the final heads.

File: src/Network.cpp [hunk 3]
```diff
@@ -30,6 +30,7 @@
 #include <boost/utility.hpp>
 #include <boost/format.hpp>
 #include <boost/spirit/home/x3.hpp>
+// #include <ctime>
```
[the remark this answers is not in the document]
Comment (OmnipotentEntity) [reply src/Network.cpp]: I commented out the timing code, so that other people could replicate my findings.  I'll remove it entirely if need be.

**Comment by OmnipotentEntity** [comment]
A kind person allowed me to use his computer to set up a match.  So far the result after 120 games is exactly 60-60.

```
leelaz-next-3 v leelaz-next-1 (120/400 games)
board size: 19   komi: 7.5
                wins              black         white       avg cpu
leelaz-next-3     60 50.00%       24 40.00%     36 60.00%    337.32
leelaz-next-1     60 50.00%       24 40.00%     36 60.00%    327.70
                                  48 40.00%     72 60.00%
```

**Comment by gcp** [comment]
Quick thoughts:

a) We're not sure what the best float format is, see "half" vs "Bfloat16" discussion. Ironically the text format can deal with both. Would be nice to have some flexibility there is we have a binary one.

b) Without working training code we can't deploy such nets.

c) If we are revising the network format, we should add beta and gamma to the batchnorm layers. We currently encode the beta into the convolution bias of the layer before, but we can't encode gamma.

**Comment by gcp** [comment]
> May I get a review? Is the file format sane and fully featured? Does the code make sense and is it readable? Is this something that the project is interested in?

I hope I sufficiently addressed this. We can continue in an issue if you want.

Improving the network format is something we should do, but it has a large impact as it needs changes all over, so I'd prefer to do it right or not at all.

**Comment by gcp** [comment]
Whethe process_bn_var should be done at writing time is another thing to consider.
